### ISSUE TITLE
Iterator.grouped leaks memory

### ISSUE THREAD

**Issue by lbialy** [javaslang/javaslang#1899]
Please consider following snippets and try to execute them:

```java
import javaslang.collection.Iterator;

class InfiniteGroupedIterators {

  private double count = 0;

  private class Message {
    private final double count;

    Message(double count) {
      this.count = count;
    }
  }

  InfiniteGroupedIterators() {
    getInfiniteIterator()
        .grouped(100)
        .zipWithIndex()
        .forEach(t -> System.out.println(String.format("Processed %s group", String.valueOf(t._2))));
  }

  private Iterator<Message> getInfiniteIterator() {
    return Iterator.continually(() -> {
      count++;
      return new Message(count);
    });
  }
}
```

vs

```scala
object InfiniteGroupedIterators {

  case class Message(count: Double)

  var index = 0d

  def infiniteIterator = Iterator.continually(() => {
    index = index + 1
    Message(index)
  })

  infiniteIterator
    .grouped(100)
    .zipWithIndex
    .foreach { t => println(s"Processed ${t._2} group") }

}
```

Javaslang's `Iterator.grouped` OOMs, Scala's `Iterator.grouped` doesn't. Used sliding view should be garbage collectable and evaluated elements shouldn't be held indefinitely, which Stream class does [here](https://github.com/javaslang/javaslang/blob/master/javaslang/src/main/java/javaslang/collection/Iterator.java#L1836) I think.

**Comment by danieldietrich** [comment]
Great finding, thank you. Will take a look this evening.

**Comment by danieldietrich** [comment]
I've shortened the Javaslang example a bit and can reproduce the error.

```java
import static javaslang.API.*;
import javaslang.collection.Iterator;

class InfiniteGroupedIterators {

    static class Message {
        final double count;
        Message(double count) { this.count = count; }
    }

    double count = 0;

    Iterator<Message> getInfiniteIterator() {
        return Iterator.continually(() -> new Message(++count));
    }

    {
        getInfiniteIterator()
                .grouped(100)
                .zipWithIndex()
                .forEach(t -> println(String.format("Processed %s group", String.valueOf(t._2))));
    }
}
```

When starting the test with -Xmx64m I receive the following output:

```java
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
	at javaslang.collection.Stream.cons(Stream.java:232)
	at javaslang.collection.StreamModule$StreamFactory.create(Stream.java:2036)
	at javaslang.collection.StreamModule$StreamFactory.lambda$create$0(Stream.java:2036)
	at javaslang.collection.StreamModule$StreamFactory$$Lambda$2/189568618.get(Unknown Source)
	at javaslang.Lazy.computeValue(Lazy.java:148)
	at javaslang.Lazy.get(Lazy.java:143)
	at javaslang.collection.StreamModule$ConsImpl.tail(Stream.java:1842)
	at javaslang.collection.Stream.drop(Stream.java:928)
	at javaslang.collection.Stream.splitAt(Stream.java:1407)
	at javaslang.collection.Iterator$41.hasNext(Iterator.java:1844)
	at javaslang.collection.Iterator$28.hasNext(Iterator.java:1148)
	at javaslang.Value.forEach(Value.java:316)
	at javaslang.InfiniteGroupedIterators.<init>(InfiniteGroupedIterators.java:35)
	at javaslang.InfiniteGroupedIterators.main(InfiniteGroupedIterators.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

I will dig deeper tomorrow...

**Comment by danieldietrich** [comment]
@lbialy Thank you for reporting the bug! We where able to fix it and speeded it up.

**Comment by lbialy** [comment]
Great, thanks a lot! Will this fix be included in next patch version or *2.1.0*?

**Comment by danieldietrich** [comment]
There will be a 2.0.6 the next days.
